pulsator4droid/src/main/java/pl/bclogic/pulsator4droid/library/circle/CirclePulsatorLayout.java

```diff
@@ -0,0 +1,392 @@
+        TypedArray attr = context.getTheme().obtainStyledAttributes(
+                attrs, R.styleable.Pulsator4Droid, 0, 0);
+
+        mRepeat = DEFAULT_REPEAT;
+        mStartFromScratch = DEFAULT_START_FROM_SCRATCH;
+        mInterpolator = DEFAULT_INTERPOLATOR;
+
+        try {
+            mRepeat = attr.getInteger(R.styleable.Pulsator4Droid_pulse_repeat, DEFAULT_REPEAT);
+            mStartFromScratch = attr.getBoolean(R.styleable.Pulsator4Droid_pulse_startFromScratch,
+                    DEFAULT_START_FROM_SCRATCH);
+            mInterpolator = attr.getInteger(R.styleable.Pulsator4Droid_pulse_interpolator,
+                    DEFAULT_INTERPOLATOR);
+
+        } finally {
+            attr.recycle();
+        }
+
+        // create views
+        build();
+    }
+
+    @Override
+    protected PulseShape getPulseShape(Paint paint) {
+        return new PulseCircle(paint);
+    }
+
+    @Override
+    protected void onDraw(Canvas canvas) {
+        if (mMask == null && mCircularMaskRadius > 0) {
+            mMask = new Path();
+            PulseCircle circle = (PulseCircle) mPulseShape;
+            mMask.addCircle(
+                    circle.mCenterX, circle.mCenterY, mCircularMaskRadius, Direction.CW);
+        }
+        if (mMask != null) {
+            canvas.clipPath(mMask, Op.DIFFERENCE);
+        }
+    }
+
+    @Override
+    protected void onDetachedFromWindow() {
+        super.onDetachedFromWindow();
+
+        if (mAnimators != null) {
+            for (Animator animator : mAnimators) {
+                animator.cancel();
+            }
+            mAnimators = null;
+        }
+    }
+
+    @Override
+    public void start() {
+
+        if (mAnimators == null) {
+            reset();
+            if (mAnimators == null) {
+                return;
+            }
+        }
+
+        if (mIsStarted) {
+            return;
+        }
+
+        for (int x = 0; x < mAnimators.size(); x++) {
+            ObjectAnimator objectAnimator = (ObjectAnimator) mAnimators.get(x);
+
+            if (!mStartFromScratch) {
+                // instead of delaying the animation, fast-forward it
+                long delay = objectAnimator.getStartDelay();
+                objectAnimator.setStartDelay(0);
+
+                // This is where it starts to get tricky. The documentation of
+                // ValueAnimator#setCurrentPlayTime(long) is a bit confusing about whether it
+                // should be called before or after starting the animation itself. The truth is,
+                // it seems the behavior differs between Android versions. If it gets called at
+                // a wrong time, only some of the animations will start while others will not or
+                // the animated object won't be visible at all.
+                boolean shouldStartBeforeSettingCurrentTime =
+                        Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP_MR1;
+
+                if (shouldStartBeforeSettingCurrentTime) {
```

Comment:
I'd personally prefer delegating this to methods:
if(shouldStartBeforeSettingCurrentTime){
start();
setCurrentPlayTime()
} else{
setCurrentPlayTime();
start;
}

It would be a bit cleaner and easier to understand.

Reply:
This whole class is basically just moved code from the original library. I would rather touch this as little as possible due to any potential updates we might do from the original repo.